README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# asymptor <img src="man/figures/logo.gif" align="right" alt="" width="120" />

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version-ago/asymptor)](https://CRAN.R-project.org/package=asymptor)
[![R build
status](https://github.com/Bisaloo/asymptor/workflows/R-CMD-check/badge.svg)](https://github.com/Bisaloo/asymptor/actions)
[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://lifecycle.r-lib.org/articles/stages.html#maturing)
[![Codecov test
coverage](https://codecov.io/gh/Bisaloo/asymptor/branch/main/graph/badge.svg)](https://app.codecov.io/gh/Bisaloo/asymptor?branch=main)
<!-- badges: end -->

The asymptor R package allows you to estimate the lower and upper bound
of asymptomatic cases in an epidemic using the capture/recapture methods
from [Böhning et al. (2020)](https://doi.org/10.1016/j.ijid.2020.06.009)
and [Rocchetti et
al. (2020)](https://doi.org/10.1101/2020.07.14.20153445).

**Please note there is currently some discussion about the validity of
the methods implemented in this package. You should read carefully the
original articles, alongside this answer from [Li et al.
(2022)](https://doi.org/10.48550/arXiv.2209.11334) before using this
package in your project.**

## Installation

You can install the stable version of this package from CRAN:

``` r
install.packages("asymptor")
```

or the development version from [GitHub](https://github.com/bisaloo),
via my [r-universe](https://bisaloo.r-universe.dev/):

``` r
install.packages("asymptor", repos = "https://bisaloo.r-universe.dev")
```

## Example

Let’s start by loading some example data from the COVID-19 epidemic in
Italy:

``` r
d <- readRDS(system.file("extdata", "covid19_italy.rds", package = "asymptor"))
head(d)
#>         date new_cases new_deaths
#> 1 2020-01-02         0          0
#> 2 2020-01-03         0          0
#> 3 2020-01-04         0          0
#> 4 2020-01-05         0          0
#> 5 2020-01-06         0          0
#> 6 2020-01-07         0          0
```

We can estimate the lower and upper bound of asymptomatic cases with:

``` r
library(asymptor)
estimate_asympto(d$date, d$new_cases, d$new_deaths)
```

Or, with a tidyverse-compatible syntax:

``` r
library(dplyr)
d %>%
  mutate(asympto_cases = estimate_asympto(date, new_cases, new_deaths))
```

Please refer to the
[vignette](https://hugogruson.fr/asymptor/articles/example.html) for a
detailed example using the COVID-19 data from Italy.

![example\_figure](https://hugogruson.fr/asymptor/articles/example_files/figure-html/example_fig-1.png)


NEWS.md
# asymptor (development version)

# asymptor 1.1.0

## Minor changes

* there is now a warning about a possible flaw in the methods in the README, the DESCRIPTION, and the `estimate_asympto()` function documentation.

# asymptor 1.0

## Major breaking changes

* the public interface of `estimate_asympto()` has changed to take three 
arguments `date`, `cases`, `deaths` instead of one `df` with three columns.
The reason for this change is that the old behaviour forced specific column
names that might not match the datasets from users.

## Minor changes

* Unexported functions now use the `@noRd` roxygen tag, which mean they will
not appear in the manual
* The citation for the package now refers to the correct package

# asymptor 0.2

* First release.


tests/testthat.R
library(testthat)
library(asymptor)

test_check("asymptor")


tests/testthat/test-estimate_asympto.R
test_that("test-format", {

  d <- readRDS(system.file("extdata", "covid19_italy.rds", package = "asymptor"))

  a_up <- estimate_asympto(d$date, d$new_cases, d$new_deaths, bounds = "upper")
  a_lw <- estimate_asympto(d$date, d$new_cases, d$new_deaths, bounds = "lower")
  a_both <- estimate_asympto(d$date, d$new_cases, d$new_deaths)

  expect_identical(dim(a_up), c(nrow(d), 2L))
  expect_identical(dim(a_lw), c(nrow(d), 2L))
  expect_identical(dim(a_both), c(nrow(d), 3L))

  expect_identical(merge(a_lw, a_up), a_both)

  expect_false(any(a_both[, c(2,3)]<0, na.rm = TRUE))

})

test_that("test-values", {

  # Austria data from https://doi.org/fbwv
  d <- readRDS(system.file("extdata", "data_fbwv.rds", package = "asymptor"))
  new_cases <- c(NA_real_, diff(d$total_cases))
  new_deaths <- c(NA_real_, diff(d$total_deaths))

  res <- estimate_asympto(d$date, new_cases, new_deaths, "lower")

  expect_identical(sum(res$lower, na.rm = TRUE), 17264L)

  # Match upper bound result from https://doi.org/10.1101/2020.07.14.20153445
  # FIXME: this doesn't work but probably because the data changed in the meantime

  # d <- readRDS(system.file("extdata", "covid19_germany.rds", package = "asymptor"))
  #
  # res <- estimate_asympto(d)
  #
  # expect_identical(
  #   sum(res$upper[res$date <= "2020-04-17"], na.rm = TRUE),
  #   650841L-137439L
  # )

})

test_that("test-error", {

  d <- readRDS(system.file("extdata", "covid19_italy.rds", package = "asymptor"))

  expect_error(
    estimate_asympto(d$date[-1], d$new_cases, d$new_deaths),
    "same lengths"
  )

})


R/estimate_asympto.R
#' Estimate the proportion of asymptomatic cases by capture/recapture
#'
#' @param date A vector containing the dates
#' @param cases A numeric vector containing the number of new cases at each
#' `date` (**not** the cumulative number of cases).
#' @param deaths A numeric vector containing the number of new deaths at each
#' `date` (**not** the cumulative number of cases).
#' @param bounds `"lower"`, `"upper"`, or both `c("lower, "upper"`) (the
#'   default), telling which bounds of the number of asymptomatic cases are
#'   computed.
# @param sd If `TRUE` (defaults to `FALSE`), returns the standard deviation
#'   associated with the lower and upper bounds.
# @param unbiased If `TRUE` (the default), returns the unbiased version of the
#'   estimation. `FALSE` is never the recommended setting excepted for testing
#'   or debugging purpose.
#'
#' @return A `data.frame` with two or three columns (depending on the value of
#'   the `bounds` argument):
#'   * `date`: the original `date` column
#'   * `lower`: the lower bound of asymptomatic cases
#'   * `upper`: the upper bound of asymptomatic cases
#'
#' @export
#'
#' @references Böhning D., Rocchetti I., Maruotti A., Holling H. (2020),
#'   Estimating the undetected infections in the Covid-19 outbreak by harnessing
#'   capture–recapture methods, International Journal of Infectious Diseases,
#'   97, p197-201, \doi{10.1016/j.ijid.2020.06.009}.
#' @references Rocchetti I., Böhning D., Holling H., Maruotti A., (2020),
#'   Estimating the size of undetected cases of the SARS-CoV-2 outbreak in
#'   Europe: An upperbound estimator, medRxiv,
#'   \doi{10.1101/2020.07.14.20153445}.
#' @references Li M., Dushoff J., Earn D.J.D., Bolker B.M. (2022), Evaluating
#'   undercounts in epidemics: responsi to Maruotti et al. 2022, arXiv,
#'   \doi{10.48550/ARXIV.2209.11334}.
#'
#' @note
#' Please note there is currently some discussion about the validity of the
#' methods implemented in this package. You should read carefully the original
#' articles, alongside this answer from Li et al. (2022) before using this
#' package in your project.
#'
#' @examples
#'
#' d <- readRDS(system.file("extdata", "covid19_italy.rds", package = "asymptor"))
#' head(d)
#'
#' estimate_asympto(d$date, d$new_cases, d$new_deaths)
#'
estimate_asympto <- function(date, cases, deaths, bounds = c("lower", "upper")) {

  bounds <- match.arg(bounds, several.ok = TRUE)

  nobs <- length(date)

  if (length(cases) != nobs || length(cases) != nobs) {
    stop(
      "The vectors passed to `date`, `cases`, and `deaths` must have the same ",
      "lengths", call. = FALSE
    )
  }

  res <- as.data.frame(date)

  f <- function(k) {
    if (k == 1) {
      cases
    } else {
      custom_lag(cases, k-1) - rowSums(vapply(0:(k-2), function(i) custom_lag(deaths, i), numeric(nobs)))
    }
  }

  f1 <- f(1)
  f2 <- f(2)

  # f0 is also the lower bound
  f0 <- f1*(f1-1)/(1+pmax(0, f2))

  if ("lower" %in% bounds) {
    res <- cbind(res, as.integer(round(f0)))
  }

  if ("upper" %in% bounds) {
#    current <- cumsum(with(df, new_cases-new_deaths-new_recoveries))
    current <- f1+f2+f(3)

    p <- function(k) {
      rowSums(vapply(seq_len(k), f, numeric(nobs))) / current
    }

    pi0 <- f0 / (f1+f2+f0)

    pi <- function(k) {
      pi0 + (1-pi0) * p(k)
    }

    pi0_ub <- (p(2)-p(1))/(1-pi(1)/pi(2)+p(2)-p(1))

    upb <- current * pi0_ub / (1-pi0_ub)

    if (any(f0 < 0, upb < 0, na.rm = TRUE)) {
      warning(
        "Number of asymptomatic cases is negative for some dates. ",
        "Manually setting to 0", call. = FALSE
      )
      f0[f0 < 0] <- 0
      upb[upb < 0] <- 0
    }

    res <- cbind(res, as.integer(round(upb)))
  }

  colnames(res) <- c("date", bounds)

  return(res)

}
